packages/jupyter-ai/jupyter_ai/history.py

```diff
@@ -30,15 +33,30 @@ async def aget_messages(self) -> List[BaseMessage]:
 
     def add_message(self, message: BaseMessage) -> None:
         """Add a self-created message to the store"""
+        if HUMAN_MSG_ID_KEY not in message.additional_kwargs:
+            # human message id must be added to allow for targeted clearing of messages.
+            # `WrappedBoundedChatHistory` should be used instead to add messages.
+            raise ValueError(
+                "Message must have a human message ID to be added to the store."
+            )
         self._all_messages.append(message)
 
     async def aadd_messages(self, messages: Sequence[BaseMessage]) -> None:
         """Add messages to the store"""
         self.add_messages(messages)
 
-    def clear(self) -> None:
-        self._all_messages = []
+    def clear(self, human_msg_id: Optional[str] = None) -> None:
+        """Clear all messages after the given time"""
+        if human_msg_id:
+            self._all_messages = [
+                m
+                for m in self._all_messages
+                if m.additional_kwargs[HUMAN_MSG_ID_KEY] != human_msg_id
+            ]
+        else:
+            self._all_messages = []
         self.clear_time = time.time()
+        self.clear_msg_id = human_msg_id
```

Comment:
1. This class doesn't need to keep track of `clear_time` anymore, since that is not used to determine if an agent response should be stored in the memory. That logic should be removed. 

2. However, this class should track a **set** of cleared messages, not just the last cleared message. For example, if I send two messages A and B (in that order), then clear A and B before receiving a response to either message (in that order), both responses should not be added to memory. If only one cleared message ID is stored, then it is possible for `self.clear_msg_id` to be set to `B.id` before A's response arrives, which causes A's response to be added to memory in `add_message()`.

So the attributes


```py
    clear_time: float = 0.0
    clear_msg_id: Optional[str] = None
```

should be replaced by:

```py
    cleared_msgs: set[str] = set()
```

While this set does technically grow infinitely as users clear more individual exchanges, it does so at such a slow rate that it should never be a problem for users. Also, if `human_msg_id` is not passed (e.g. `/clear` or `New chat` is called), we can clear that set from memory.

Reply:
Also, can you update the docstring here as well?

Reply:
Good catch on requiring a list of ids.

`clear_time` is required for full clears. Its purpose is to ensure if clear is triggered while a response is pending, this pending exchange is not added to the chat memory. E.g. A < Clear < Ar, exchange of (A, Ar) would be received at Ar, after Clear, and would be sent to `add_message`. So condition human_msg.time > clear_time is checked before adding. This was originally the entire purpose of the wrapped class.